most builds seem to be stalling and losing connection

We're seeing many of the builds failing. 

I've noticed a trend that it seems to slow down and then stall the rendering the index pages. It's not reliably at the same place but it does appear to be in the same region. I've upped the timeout to 3 hours and it's still seeing the same behavior. My next area to look will be at resource starvation, in particular memory. The slowdown could be swappyness before the OOM killer takes down the jenkins agent. 

Note that when it crashes we're not near the end of the list. There's 114 pages of package listings. And the successful builds finish in 1:20. The latest builds I upped the timeout to 3 hours and that does not appear to be the issue.

http://build.ros.org/job/doc_rosindex/359/console
<details>
01:48:46.989          Rendering: packages/page/50/name/index.html
01:48:46.989   Pre-Render Hooks: packages/page/50/name/index.html
01:48:46.989   Rendering Liquid: packages/page/50/name/index.html
01:48:47.179   Rendering Markup: packages/page/50/name/index.html
01:48:47.179   Rendering Layout: packages/page/50/name/index.html
01:48:47.179      Layout source: site
01:48:47.278          Rendering: packages/page/50/time/index.html
01:48:47.278   Pre-Render Hooks: packages/page/50/time/index.html
01:48:47.278   Rendering Liquid: packages/page/50/time/index.html
01:48:47.431   Rendering Markup: packages/page/50/time/index.html
01:48:47.431   Rendering Layout: packages/page/50/time/index.html
01:48:47.431      Layout source: site
01:48:47.785          Rendering: packages/page/50/doc/index.html
01:48:47.785   Pre-Render Hooks: packages/page/50/doc/index.html
01:48:47.785   Rendering Liquid: packages/page/50/doc/index.html
01:48:48.892   Rendering Markup: packages/page/50/doc/index.html
01:48:48.892   Rendering Layout: packages/page/50/doc/index.html
01:48:48.892      Layout source: site
01:48:49.105          Rendering: packages/page/50/released/index.html
01:48:49.105   Pre-Render Hooks: packages/page/50/released/index.html
01:48:49.105   Rendering Liquid: packages/page/50/released/index.html
01:48:50.119   Rendering Markup: packages/page/50/released/index.html
01:48:50.119   Rendering Layout: packages/page/50/released/index.html
01:48:50.119      Layout source: site
01:49:05.222          Rendering: packages/page/51/name/index.html
01:49:56.590   Pre-Render Hooks: packages/page/51/name/index.html
01:50:02.630   Rendering Liquid: packages/page/51/name/index.html
02:00:00.423 FATAL: command execution failed
02:00:00.423 java.nio.channels.ClosedChannelException
</details>

http://build.ros.org/job/doc_rosindex/358/console
<details>
01:58:01.173          Rendering: packages/page/24/released/index.html
01:58:01.173   Pre-Render Hooks: packages/page/24/released/index.html
01:58:01.173   Rendering Liquid: packages/page/24/released/index.html
01:58:01.327   Rendering Markup: packages/page/24/released/index.html
01:58:01.327   Rendering Layout: packages/page/24/released/index.html
01:58:01.327      Layout source: site
01:58:01.440          Rendering: packages/page/25/name/index.html
01:58:01.440   Pre-Render Hooks: packages/page/25/name/index.html
01:58:01.440   Rendering Liquid: packages/page/25/name/index.html
01:58:01.560   Rendering Markup: packages/page/25/name/index.html
01:58:01.560   Rendering Layout: packages/page/25/name/index.html
01:58:01.560      Layout source: site
01:58:01.719          Rendering: packages/page/25/time/index.html
01:58:01.719   Pre-Render Hooks: packages/page/25/time/index.html
01:58:01.719   Rendering Liquid: packages/page/25/time/index.html
01:58:01.991   Rendering Markup: packages/page/25/time/index.html
01:58:01.991   Rendering Layout: packages/page/25/time/index.html
01:58:01.991      Layout source: site
01:58:03.314          Rendering: packages/page/25/doc/index.html
01:58:03.314   Pre-Render Hooks: packages/page/25/doc/index.html
01:58:03.314   Rendering Liquid: packages/page/25/doc/index.html
01:58:03.727   Rendering Markup: packages/page/25/doc/index.html
01:58:03.727   Rendering Layout: packages/page/25/doc/index.html
01:58:03.727      Layout source: site
01:58:08.382          Rendering: packages/page/25/released/index.html
01:58:12.519   Pre-Render Hooks: packages/page/25/released/index.html
01:58:15.134   Rendering Liquid: packages/page/25/released/index.html
02:00:09.948   Rendering Markup: packages/page/25/released/index.html
02:00:09.948   Rendering Layout: packages/page/25/released/index.html
02:00:09.948      Layout source: site
02:00:23.703          Rendering: packages/page/26/name/index.html
02:00:47.014   Pre-Render Hooks: packages/page/26/name/index.html
02:00:53.738   Rendering Liquid: packages/page/26/name/index.html
02:05:14.485 FATAL: command execution failed
02:05:14.486 java.nio.channels.ClosedChannelException
</details>

There was an additional one recetly which make was "Killed" after the slow down.

http://build.ros.org/job/doc_rosindex/365/console
<details>
01:57:27.746   Pre-Render Hooks: p/pr2_plugs_msgs/index.html
01:57:27.746   Rendering Liquid: p/pr2_plugs_msgs/index.html
01:57:30.686   Rendering Markup: p/pr2_plugs_msgs/index.html
01:57:30.864   Rendering Layout: p/pr2_plugs_msgs/index.html
01:57:31.451      Layout source: site
01:59:32.039 make: *** [build] Killed
01:59:32.049 Makefile:19: recipe for target 'build' failed
</details>

--- a/_layouts/deps.html
+++ b/_layouts/deps.html
@@ -31,13 +31,6 @@ layout: default
                 {% endif %}
               </th>
               <th style="text-align: center;">
-                {% if page.sort_id == 'doc' %}
-                  <span class="glyphicon glyphicon-file selected-sort-label"></span>
-                {% else %}
-                  <a href="{{site.baseurl}}/packages/page/1/doc" class="unselected-sort-label glyphicon glyphicon-file"></a>
-                {% endif %}
-              </th>
-              <th style="text-align: center;">
                 {% if page.sort_id == 'time' %}
                   <span class="glyphicon glyphicon-time selected-sort-label"></span>
                 {% else %}
@@ -72,15 +65,6 @@ layout: default
                     </td>
                   {% else %}
                     <td title="Not released in {{distro}}" data-toggle="tooltip" data-placement="left" class="text-center">
-                      <span class="glyphicon glyphicon-none"></span>
-                    </td>
-                  {% endif %}
-                  {% if p.data.readme %}
-                    <td title="Has README" data-toggle="tooltip" data-placement="left" align="center">
-                      <span class="glyphicon glyphicon-file"></span>
-                    </td>
-                  {% else %}
-                    <td title="No README" data-toggle="tooltip" data-placement="left" align="center">
                       <span class="glyphicon glyphicon-none"></span>
                     </td>
                   {% endif %}

--- a/_layouts/packages.html
+++ b/_layouts/packages.html
@@ -39,16 +39,6 @@ layout: default
                     {% endif %}
                   </th>
                   <th style="text-align: center;">
-                    {% if page.sort_id == 'doc' %}
-                      <span class="glyphicon glyphicon-file selected-sort-label"
-                            title="Sorted by README status"></span>
-                    {% else %}
-                      <a href="{{site.baseurl}}/packages/page/1/doc"
-                         class="unselected-sort-label glyphicon glyphicon-file"
-                         title="Sort by README status"></a>
-                    {% endif %}
-                  </th>
-                  <th style="text-align: center;">
                     {% if page.sort_id == 'time' %}
                       <span class="glyphicon glyphicon-time selected-sort-label"
                             title="Sorted by last commit date"></span>
@@ -89,16 +79,6 @@ layout: default
                       </td>
                     {% else %}
                       <td title="Not released in {{distro}}" data-toggle="tooltip" data-placement="left" class="text-center">
-                        <span class="glyphicon glyphicon-none"></span>
-                      </td>
-                    {% endif %}
-                    {% assign n_readmes = p.data[1].readmes | size %}
-                    {% if n_readmes > 0 %}
-                      <td title="Has README" data-toggle="tooltip" data-placement="left" align="center">
-                        <span class="glyphicon glyphicon-file"></span>
-                      </td>
-                    {% else %}
-                      <td title="No README" data-toggle="tooltip" data-placement="left" align="center">
                         <span class="glyphicon glyphicon-none"></span>
                       </td>
                     {% endif %}

--- a/_layouts/repos.html
+++ b/_layouts/repos.html
@@ -40,16 +40,6 @@ layout: default
                     {% endif %}
                   </th>
                   <th style="text-align: center;">
-                    {% if page.sort_id == 'doc' %}
-                      <span class="glyphicon glyphicon-file selected-sort-label"
-                            title="Sorted by README status"></span>
-                    {% else %}
-                      <a href="{{site.baseurl}}/repos/page/1/doc"
-                         class="unselected-sort-label glyphicon glyphicon-file"
-                         title="Sort by README status"></a>
-                    {% endif %}
-                  </th>
-                  <th style="text-align: center;">
                     {% if page.sort_id == 'time' %}
                       <span class="glyphicon glyphicon-time selected-sort-label"
                             title="Sorted by last commit date"></span>
@@ -87,15 +77,6 @@ layout: default
                       </td>
                     {% else %}
                       <td title="Not released in {{distro}}" data-toggle="tooltip" data-placement="left" class="text-center">
-                        <span class="glyphicon glyphicon-none"></span>
-                      </td>
-                    {% endif %}
-                    {% if r.data.readme %}
-                      <td title="Has README" data-toggle="tooltip" data-placement="left" align="center">
-                        <span class="glyphicon glyphicon-file"></span>
-                      </td>
-                    {% else %}
-                      <td title="No README" data-toggle="tooltip" data-placement="left" align="center">
                         <span class="glyphicon glyphicon-none"></span>
                       </td>
                     {% endif %}

--- a/_plugins/rosindex_generator.rb
+++ b/_plugins/rosindex_generator.rb
@@ -681,7 +681,7 @@ class Indexer < Jekyll::Generator
   end
 
   def sort_repos(site)
-    repos_sorted = {'name' => {}, 'time' => {}, 'doc' => {}, 'released' => {}}
+    repos_sorted = {'name' => {}, 'time' => {}, 'released' => {}}
 
     repos_sorted_by_name = @repo_names.sort_by { |name, _| name }
     $all_distros.collect do |distro|
@@ -696,13 +696,6 @@ class Indexer < Jekyll::Generator
         end.max.to_s
       end.reverse
 
-      repos_sorted['doc'][distro] = \
-      repos_sorted['name'][distro].sort_by do |_, instances|
-        instances.default.snapshots.count do |d, s|
-          d == distro and not s.nil? and not s.data['readme'].nil?
-        end
-      end.reverse
-
       repos_sorted['released'][distro] = \
       repos_sorted['name'][distro].sort_by do |_, instances|
         instances.default.snapshots.count do |d, s|
@@ -715,7 +708,7 @@ class Indexer < Jekyll::Generator
   end
 
   def sort_packages(site)
-    packages_sorted = {'name' => {}, 'time' => {}, 'doc' => {}, 'released' => {}}
+    packages_sorted = {'name' => {}, 'time' => {}, 'released' => {}}
 
     packages_sorted_by_name = @package_names.sort_by { |name, _| name }
     $all_distros.each do |distro|
@@ -728,13 +721,6 @@ class Indexer < Jekyll::Generator
         end.map do |_, s|
           s.snapshot.data['last_commit_time'].to_s
         end.max.to_s
-      end.reverse
-
-      packages_sorted['doc'][distro] = \
-      packages_sorted['name'][distro].sort_by do |_, instances|
-        instances.snapshots.count do |d, s|
-          distro == d and not s.nil? and s.data['readmes'].count > 0
-        end
       end.reverse
 
       packages_sorted['released'][distro] = \

